File: runtimes/node/src/agent.ts
import * as net from 'net';

export class AgenticStream {
    private callId = 0;
    private buffer = '';
    private readonly pendingResolvers: Map<number, (res: any) => void> = new Map();

    constructor(private readonly socket: net.Socket) {
        this.socket.on('data', (chunk) => {
            this.buffer += chunk.toString('utf-8');
            let boundary = this.buffer.indexOf('\n');
            while (boundary !== -1) {
                const line = this.buffer.slice(0, boundary).trim();
                this.buffer = this.buffer.slice(boundary + 1);
                if (line) {
                    try {
                        const frame = JSON.parse(line);
                        if (frame.type === 'tool_response') {
                            const resolver = this.pendingResolvers.get(frame.id);
                            if (resolver) {
                                resolver(frame.result || {});
                                this.pendingResolvers.delete(frame.id);
                            }
                        }
                    } catch (e) {
                        // ignore malformed JSON
                    }
                }
                boundary = this.buffer.indexOf('\n');
            }
        });
    }

    public sendToken(token: string): Promise<void> {
        return new Promise((resolve, reject) => {
            const payload = JSON.stringify({ type: 'token', content: token }) + '\n';
            this.socket.write(payload, 'utf-8', (err) => {
                if (err) reject(err);
                else resolve();
            });
        });
    }

    public callTool(toolName: string, args: Record<string, any>): Promise<Record<string, any>> {
        this.callId++;
        const currentId = this.callId;

        return new Promise((resolve, reject) => {
            this.pendingResolvers.set(currentId, resolve);

            const payload = JSON.stringify({
                type: 'tool_call',
                id: currentId,
                name: toolName,
                arguments: args
            }) + '\n';

            this.socket.write(payload, 'utf-8', (err) => {
                if (err) {
                    this.pendingResolvers.delete(currentId);
                    reject(err);
                }
            });
        });
    }
}

File: runtimes/python/helix_rt/agent.py
import asyncio
import json


class AgenticStream:
    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        self.reader = reader
        self.writer = writer
        self.call_id = 0

    async def send_token(self, token: str) -> None:
        """Sends a text token to the client."""
        payload = json.dumps({"type": "token", "content": token}) + "\n"
        self.writer.write(payload.encode("utf-8"))
        await self.writer.drain()

    async def call_tool(self, tool_name: str, args: dict) -> dict:
        """Invokes a client-side tool, suspends execution, and blocks waiting for the response."""
        self.call_id += 1
        current_id = self.call_id
        
        # Write tool call request frame
        payload = json.dumps({
            "type": "tool_call",
            "id": current_id,
            "name": tool_name,
            "arguments": args
        }) + "\n"
        self.writer.write(payload.encode("utf-8"))
        await self.writer.drain()

        # Block reading lines until we find the tool_response matching current_id
        while True:
            line = await self.reader.readline()
            if not line:
                raise EOFError("connection closed while waiting for tool response")
            
            try:
                frame = json.loads(line.decode("utf-8").strip())
                if frame.get("type") == "tool_response" and frame.get("id") == current_id:
                    return frame.get("result", {})
            except Exception:
                pass
